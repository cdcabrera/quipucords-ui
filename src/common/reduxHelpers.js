import helpers from './helpers';

const generatedPromiseActionReducer = (types = [], state = {}, action = {}) => {
  if (!action.meta || !action.meta.id) {
    return state;
  }

  const { type } = action;
  const [whichType] = types.filter(val =>
    new RegExp(
      `^(${helpers.REJECTED_ACTION(val.type || val)}|${helpers.PENDING_ACTION(
        val.type || val
      )}|${helpers.FULFILLED_ACTION(val.type || val)})$`
    ).test(type)
  );

  if (!whichType) {
    return state;
  }

  switch (type) {
    case helpers.REJECTED_ACTION(whichType.type || whichType):
      return helpers.setStateProp(
        whichType.ref || null,
        {
          [action.meta.id]: {
            error: action.error,
            errorMessage: helpers.getMessageFromResults(action.payload).message,
            errorStatus: helpers.getStatusFromResults(action.payload),
            metaData: action.meta.data,
            metaId: action.meta.id,
            metaQuery: action.meta.query
          }
        },
        {
          state
        }
      );
    case helpers.PENDING_ACTION(whichType.type || whichType):
      return helpers.setStateProp(
        whichType.ref || null,
        {
          [action.meta.id]: {
            metaData: action.meta.data,
            metaId: action.meta.id,
            metaQuery: action.meta.query,
            pending: true
          }
        },
        {
          state
        }
      );

    case helpers.FULFILLED_ACTION(whichType.type || whichType):
      return helpers.setStateProp(
        whichType.ref || null,
        {
          [action.meta.id]: {
            data: (action.payload && action.payload.data) || {},
            fulfilled: true,
            metaData: action.meta.data,
            metaId: action.meta.id,
            metaQuery: action.meta.query
          }
        },
        {
          state
        }
      );

    default:
      return state;
  }
};

const reduxHelpers = {
  generatedPromiseActionReducer
};

export { reduxHelpers as default, reduxHelpers, generatedPromiseActionReducer };
