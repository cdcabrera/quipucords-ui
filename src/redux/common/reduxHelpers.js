import { helpers } from '../../common/helpers';

const FULFILLED_ACTION = base => `${base}_FULFILLED`;

const PENDING_ACTION = base => `${base}_PENDING`;

const REJECTED_ACTION = base => `${base}_REJECTED`;

const setStateProp = (prop, data, options) => {
  const { state = {}, initialState = {}, reset = true } = options;
  let obj = { ...state };

  if (process.env.REACT_APP_ENV === 'development' && prop && !state[prop]) {
    console.error(`Error: Property ${prop} does not exist within the passed state.`, state);
  }

  if (process.env.REACT_APP_ENV === 'development' && reset && prop && !initialState[prop]) {
    console.warn(`Warning: Property ${prop} does not exist within the passed initialState.`, initialState);
  }

  if (reset && prop) {
    obj[prop] = {
      ...state[prop],
      ...initialState[prop],
      ...data
    };
  } else if (reset && !prop) {
    obj = {
      ...state,
      ...initialState,
      ...data
    };
  } else if (prop) {
    obj[prop] = {
      ...state[prop],
      ...data
    };
  } else {
    obj = {
      ...state,
      ...data
    };
  }

  return obj;
};

const generatedPromiseActionReducer = (types = [], state = {}, action = {}) => {
  if (!action.meta || !action.meta.id) {
    return state;
  }

  const { type } = action;
  const expandedTypes = [];

  types.forEach(
    val =>
      (Array.isArray(val.type) && val.type.forEach(subVal => expandedTypes.push({ ref: val.ref, type: subVal }))) ||
      expandedTypes.push(val)
  );

  const [whichType] = expandedTypes.filter(val =>
    new RegExp(
      `^(${REJECTED_ACTION(val.type || val)}|${PENDING_ACTION(val.type || val)}|${FULFILLED_ACTION(val.type || val)})$`
    ).test(type)
  );

  if (!whichType) {
    return state;
  }

  switch (type) {
    case REJECTED_ACTION(whichType.type || whichType):
      return setStateProp(
        whichType.ref || null,
        {
          [action.meta.id]: {
            error: action.error,
            errorMessage: helpers.getMessageFromResults(action.payload).message,
            errorStatus: helpers.getStatusFromResults(action.payload),
            metaData: action.meta.data,
            metaId: action.meta.id,
            metaQuery: action.meta.query
          }
        },
        {
          state
        }
      );
    case PENDING_ACTION(whichType.type || whichType):
      return setStateProp(
        whichType.ref || null,
        {
          [action.meta.id]: {
            metaData: action.meta.data,
            metaId: action.meta.id,
            metaQuery: action.meta.query,
            pending: true
          }
        },
        {
          state
        }
      );

    case FULFILLED_ACTION(whichType.type || whichType):
      return setStateProp(
        whichType.ref || null,
        {
          [action.meta.id]: {
            data: (action.payload && action.payload.data) || {},
            fulfilled: true,
            metaData: action.meta.data,
            metaId: action.meta.id,
            metaQuery: action.meta.query
          }
        },
        {
          state
        }
      );

    default:
      return state;
  }
};

const reduxHelpers = {
  FULFILLED_ACTION,
  PENDING_ACTION,
  REJECTED_ACTION,
  generatedPromiseActionReducer,
  setStateProp
};

export { reduxHelpers as default, reduxHelpers };
