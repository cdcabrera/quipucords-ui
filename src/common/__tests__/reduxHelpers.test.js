import { reduxHelpers } from '../reduxHelpers';
import { helpers } from '../helpers';

describe('ReduxHelpers', () => {
  it('should have specific functions', () => {
    expect(reduxHelpers).toMatchSnapshot('reduxHelpers');
  });

  it('should generate a standard reducer from promise actions', () => {
    const state = {
      lorem: {},
      ipsum: {}
    };

    const action = {
      meta: {
        id: 'lorem-id'
      },
      payload: {
        data: {
          test: 'test'
        }
      }
    };

    expect(
      reduxHelpers.generatedPromiseActionReducer(
        [{ ref: 'lorem', type: 'LOREM' }, { ref: 'ipsum', type: 'IPSUM' }],
        state,
        { ...action, type: helpers.FULFILLED_ACTION('LOREM') }
      )
    ).toMatchSnapshot('generatedPromiseActionReducer fulfilled');

    expect(
      reduxHelpers.generatedPromiseActionReducer(
        [{ ref: 'lorem', type: 'LOREM' }, { ref: 'ipsum', type: 'IPSUM' }],
        state,
        { ...action, type: helpers.REJECTED_ACTION('LOREM') }
      )
    ).toMatchSnapshot('generatedPromiseActionReducer rejected');

    expect(
      reduxHelpers.generatedPromiseActionReducer(
        [{ ref: 'lorem', type: 'LOREM' }, { ref: 'ipsum', type: 'IPSUM' }],
        state,
        { ...action, type: helpers.PENDING_ACTION('LOREM') }
      )
    ).toMatchSnapshot('generatedPromiseActionReducer rejected');
  });
});
