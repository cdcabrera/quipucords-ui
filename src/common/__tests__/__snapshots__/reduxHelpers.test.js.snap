// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxHelpers should generate a standard reducer from promise actions: generatedPromiseActionReducer fulfilled 1`] = `
Object {
  "ipsum": Object {},
  "lorem": Object {
    "lorem-id": Object {
      "data": Object {
        "test": "test",
      },
      "fulfilled": true,
      "metaData": undefined,
      "metaId": "lorem-id",
      "metaQuery": undefined,
    },
  },
}
`;

exports[`ReduxHelpers should generate a standard reducer from promise actions: generatedPromiseActionReducer rejected 1`] = `
Object {
  "ipsum": Object {},
  "lorem": Object {
    "lorem-id": Object {
      "error": undefined,
      "errorMessage": " Server is currently unable to handle this request.",
      "errorStatus": 0,
      "metaData": undefined,
      "metaId": "lorem-id",
      "metaQuery": undefined,
    },
  },
}
`;

exports[`ReduxHelpers should generate a standard reducer from promise actions: generatedPromiseActionReducer rejected 2`] = `
Object {
  "ipsum": Object {},
  "lorem": Object {
    "lorem-id": Object {
      "metaData": undefined,
      "metaId": "lorem-id",
      "metaQuery": undefined,
      "pending": true,
    },
  },
}
`;

exports[`ReduxHelpers should have specific functions: reduxHelpers 1`] = `
Object {
  "generatedPromiseActionReducer": [Function],
}
`;
